servlet/src/main/java/io/undertow/servlet/spec/RequestDispatcherImpl.java

```diff
@@ -253,10 +254,13 @@ private void forwardImpl(ServletRequest request, ServletResponse response, Servl
         } finally {
             servletRequestContext.setServletRequest(oldRequest);
             servletRequestContext.setServletResponse(oldResponse);
-            requestImpl.getExchange().setRelativePath(oldPath);
-            requestImpl.getExchange().getAttachment(ServletRequestContext.ATTACHMENT_KEY).setServletPathMatch(oldServletPathMatch);
-            requestImpl.getExchange().setRequestPath(oldRequestPath);
-            requestImpl.getExchange().setRequestURI(oldURI);
+            final boolean preservePath = requestImpl.getExchange().getConnection().getUndertowOptions().get(UndertowOptions.PRESERVE_PATH_ON_FORWARD, true);
```

Comment:
This should be `false` by default. A new system property should be added as a configurable switch for users who would like to revert the behavior to the previous version of JBoss (JBoss EAP 6.x).

Because, if this is set to `true` by default, it violates the servlet spec. Please refer to the Servlet API document [HttpServletRequest#getRequestURL](https://docs.oracle.com/javaee/7/api/javax/servlet/http/HttpServletRequest.html#getRequestURL--) which says:

> If this request has been forwarded using RequestDispatcher.forward(javax.servlet.ServletRequest, javax.servlet.ServletResponse), the server path in the reconstructed URL must reflect the path used to obtain the RequestDispatcher, and not the server path specified by the client.